src/_P081_Cron.ino

```diff
@@ -176,168 +220,156 @@ boolean Plugin_081(byte function, struct EventStruct *event, String& string)
+        if ((last_exec_time > current_time) || (last_exec_time == CRON_INVALID_INSTANT)) {
+          // Last execution time cannot be correct.
+          last_exec_time = CRON_INVALID_INSTANT;
+          const time_t tmp_next = P081_computeNextCronTime(event->TaskIndex, current_time);
+
+          if ((tmp_next < next_exec_time) || (next_exec_time == CRON_INVALID_INSTANT)) {
+            next_exec_time = tmp_next;
+          }
+          P081_setCronExecTimes(event, CRON_INVALID_INSTANT, next_exec_time);
         }
-        else
-        {
-          last = current_t; // Set current time;
-          next = cron_next((cron_expr *)&expr, last);
-          if(next != CRON_INVALID_INSTANT)
-          {
-#if PLUGIN_081_DEBUG
-            serialPrint(F("LastExecution:"));
-            serialPrintln(getDateTimeString(*gmtime(&last)));
-            serialPrint(F("NextExecution:"));
-            serialPrintln(getDateTimeString(*gmtime(&next)));
-#endif
-
-            converter.time = last;
-            LASTEXECUTION = converter.value;
-            converter.time = next;
-            NEXTEXECUTION = converter.value;
-#if PLUGIN_081_DEBUG
-            serialPrintln(F("Check Conversion"));
-            serialPrintln(last);
-            converter.value = LASTEXECUTION;
-            serialPrintln(converter.time);
-            serialPrintln(next);
-            converter.value = NEXTEXECUTION;
-            serialPrintln(converter.time);
-#endif
-            addLog(LOG_LEVEL_DEBUG, String(F("Next execution:")) + getDateTimeString(*gmtime(&next)));
-            LoadTaskSettings(event->TaskIndex);
-            if(function != PLUGIN_TIME_CHANGE) {
-              eventQueue.add(String(F("Cron#")) + String(ExtraTaskSettings.TaskDeviceName));
+
+        if (next_exec_time != CRON_INVALID_INSTANT) {
+          if ((function == PLUGIN_TIME_CHANGE) || (next_exec_time <= current_time)) {
+            addLog(LOG_LEVEL_DEBUG, F("Cron Elapsed"));
+            last_exec_time = next_exec_time;
+            next_exec_time = P081_computeNextCronTime(event->TaskIndex, last_exec_time);
+            P081_setCronExecTimes(event, last_exec_time, next_exec_time);
+
+            addLog(LOG_LEVEL_DEBUG, String(F("Next execution:")) + getDateTimeString(*gmtime(&next_exec_time)));
+
+            if (function != PLUGIN_TIME_CHANGE) {
+              String pluginName = getPluginNameFromDeviceIndex(getDeviceIndex_from_TaskIndex(event->TaskIndex));
+              eventQueue.add(String(F("Cron#")) + pluginName);
+              success = true;
             }
           }
-          else
-          {
-            log = String(F("CRON: INVALID INSTANT"));
-            addLog(LOG_LEVEL_ERROR, log);
-          }
+        } else {
+          addLog(LOG_LEVEL_ERROR, F("CRON: INVALID INSTANT"));
         }
-
+      } else {
+        addLog(LOG_LEVEL_ERROR, F("CRON: Time not synced"));
       }
-      success = true;
-	  break;
+      break;
     }
 
     case PLUGIN_TEN_PER_SECOND:
```

Comment:
@TD-er why not remove this part?

Reply:
What part?

The `break;` is needed as it is part of the case statement.
And the success is set after an event is added to the event queue.

The success is not used as return value from the  `PLUGIN_ONCE_A_SECOND` call, but I thought it would make it more clear for a reader to show when we achieved what we were aiming for.

Reply:
@TD-er section PLUGIN_TEN_PER_SECOND